File: go/euler083.go
// https://projecteuler.net/problem=83

// Path Sum: Four Ways

// In the 5 by 5 matrix below, the minimal path sum from the top left to the bottom right, by moving left, right, up, and down, is indicated in bold red and is equal to 2297.

// Find the minimal path sum from the top left to the bottom right by moving left, right, up, and down in matrix.txt (right click and "Save Link/Target As..."), a 31K text file containing an 80 by 80 matrix.

// Answer: 425185

// Execution time: ~10ms

// Optimizations: Dijkstra's algorithm with priority queue

// Interesting notes: AI-generated solution using heap-based Dijkstra

package main

import (
	"bufio"
	"container/heap"
	"fmt"
	"math"
	"os"
	"strconv"
	"strings"
)

type Item struct {
	value    int
	priority int
	index    int
}

type PriorityQueue []*Item

func (pq PriorityQueue) Len() int { return len(pq) }

func (pq PriorityQueue) Less(i, j int) bool { return pq[i].priority < pq[j].priority }

func (pq PriorityQueue) Swap(i, j int) {
	pq[i], pq[j] = pq[j], pq[i]
	pq[i].index = i
	pq[j].index = j
}

func (pq *PriorityQueue) Push(x interface{}) {
	n := len(*pq)
	item := x.(*Item)
	item.index = n
	*pq = append(*pq, item)
}

func (pq *PriorityQueue) Pop() interface{} {
	old := *pq
	n := len(old)
	item := old[n-1]
	old[n-1] = nil
	item.index = -1
	*pq = old[0 : n-1]
	return item
}

func pathSumFourWays() int {
	matrix := readMatrix()
	size := len(matrix)
	dx := []int{-1, 1, 0, 0}
	dy := []int{0, 0, -1, 1}
	dist := make([][]int, size)
	for i := range dist {
		dist[i] = make([]int, size)
		for j := range dist[i] {
			dist[i][j] = math.MaxInt32
		}
	}
	dist[0][0] = matrix[0][0]
	pq := &PriorityQueue{}
	heap.Init(pq)
	heap.Push(pq, &Item{value: 0*size + 0, priority: matrix[0][0]})
	for pq.Len() > 0 {
		item := heap.Pop(pq).(*Item)
		x := item.value / size
		y := item.value % size
		if x == size-1 && y == size-1 {
			return dist[x][y]
		}
		for d := 0; d < 4; d++ {
			nx := x + dx[d]
			ny := y + dy[d]
			if nx >= 0 && nx < size && ny >= 0 && ny < size {
				if dist[nx][ny] > dist[x][y]+matrix[nx][ny] {
					dist[nx][ny] = dist[x][y] + matrix[nx][ny]
					heap.Push(pq, &Item{value: nx*size + ny, priority: dist[nx][ny]})
				}
			}
		}
	}
	return -1
}

func readMatrix() [][]int {
	file, err := os.Open("matrix.txt")
	if err != nil {
		panic(err)
	}
	defer file.Close()
	scanner := bufio.NewScanner(file)
	matrix := [][]int{}
	for scanner.Scan() {
		line := scanner.Text()
		parts := strings.Split(line, ",")
		row := []int{}
		for _, p := range parts {
			num, _ := strconv.Atoi(strings.TrimSpace(p))
			row = append(row, num)
		}
		matrix = append(matrix, row)
	}
	return matrix
}

func main() {
	fmt.Println(pathSumFourWays())
}

File: cpp/src/euler083.cpp
// https://projecteuler.net/problem=83
// Path Sum: Four Ways
//
// NOTE: This problem is a significantly more challenging version of Problem 81.
//
// In the 5 by 5 matrix below, the minimal path sum from the top left to the bottom right, by moving left, right, up, and down, is indicated in bold red and is equal to 2297.
//
// $$ \begin{pmatrix} \color{red}{131} & 673 & \color{red}{234} & \color{red}{103} & \color{red}{18}\\ \color{red}{201} & \color{red}{96} & \color{red}{342} & 965 & \color{red}{150}\\ 630 & 803 & 746 & \color{red}{422} & \color{red}{111}\\ 537 & 699 & 497 & \color{red}{121} & 956\\ 805 & 732 & 524 & \color{red}{37} & \color{red}{331} \end{pmatrix} $$
//
// Find the minimal path sum from the top left to the bottom right by moving left, right, up, and down in matrix.txt (right click and "Save Link/Target As..."), a 31K text file containing an 80 by 80 matrix.
//
// Answer: 425185

#include <iostream>
#include <vector>
#include <fstream>
#include <sstream>
#include <queue>
#include <algorithm>
#include <climits>

int path_sum_four_ways() {
    std::ifstream file("matrix.txt");
    std::vector<std::vector<int>> matrix(80, std::vector<int>(80));
    std::string line;
    int row = 0;
    while (std::getline(file, line) && row < 80) {
        std::stringstream ss(line);
        std::string token;
        int col = 0;
        while (std::getline(ss, token, ',') && col < 80) {
            matrix[row][col] = std::stoi(token);
            ++col;
        }
        ++row;
    }
    std::vector<std::vector<int>> dist(80, std::vector<int>(80, INT_MAX));
    dist[0][0] = matrix[0][0];
    std::priority_queue<std::tuple<int, int, int>, std::vector<std::tuple<int, int, int>>, std::greater<>> pq;
    pq.push({matrix[0][0], 0, 0});
    int di[4] = {-1, 0, 1, 0};
    int dj[4] = {0, 1, 0, -1};
    while (!pq.empty()) {
        auto [cost, i, j] = pq.top(); pq.pop();
        if (cost > dist[i][j]) continue;
        for (int d = 0; d < 4; ++d) {
            int ni = i + di[d], nj = j + dj[d];
            if (ni >= 0 && ni < 80 && nj >= 0 && nj < 80) {
                int ncost = cost + matrix[ni][nj];
                if (ncost < dist[ni][nj]) {
                    dist[ni][nj] = ncost;
                    pq.push({ncost, ni, nj});
                }
            }
        }
    }
    return dist[79][79];
}

#if ! defined UNITTEST_MODE
int main(int argc, char const *argv[])
{
    std::cout << "Answer: " << path_sum_four_ways() << std::endl;
}
#endif